src/cunumeric/sort/sort.cu

```diff
@@ -95,277 +137,1514 @@ struct SortPiece {
+                                 size_send.ptr(num_segments_l),
+                                 num_segments_l_aligned * sizeof(size_t),
+                                 sizeof(int64_t),
+                                 num_sort_ranks,
+                                 cudaMemcpyDeviceToHost,
+                                 stream));
+    CHECK_CUDA(cudaMemcpy2DAsync(size_recv_total.ptr(0),
+                                 1 * sizeof(size_t),
+                                 size_recv.ptr(num_segments_l),
+                                 num_segments_l_aligned * sizeof(size_t),
+                                 sizeof(int64_t),
+                                 num_sort_ranks,
+                                 cudaMemcpyDeviceToHost,
+                                 stream));
+
+    // need to sync as we share values in between host/device
+    CHECK_CUDA(cudaStreamSynchronize(stream));
   }
 
   // copy values into aligned send buffer
-  auto val_send_buf = local_sorted.values;
-  if (buf_size_send_vals_total > volume) {
-    val_send_buf = create_buffer<VAL>(buf_size_send_vals_total, Memory::GPU_FB_MEM);
-    size_t pos   = 0;
-    for (size_t r = 0; r < num_ranks; ++r) {
-      CHECK_CUDA(cudaMemcpyAsync(val_send_buf.ptr(aligned_pos_vals_send[r]),
-                                 local_sorted.values.ptr(pos),
-                                 sizeof(VAL) * size_send[r],
-                                 cudaMemcpyDeviceToDevice,
-                                 stream));
-      pos += size_send[r];
+  std::vector<Buffer<VAL>> val_send_buffers(num_sort_ranks);
+  std::vector<Buffer<int64_t>> idc_send_buffers(num_sort_ranks);
+  {
+    for (size_t r = 0; r < num_sort_ranks; r++) {
+      val_send_buffers[r] = create_buffer<VAL>(size_send_total[r], Memory::GPU_FB_MEM);
+      if (argsort) {
+        idc_send_buffers[r] = create_buffer<int64_t>(size_send_total[r], Memory::GPU_FB_MEM);
+      }
     }
-    local_sorted.values.destroy();
-  }
 
-  // copy indices into aligned send buffer
-  auto idc_send_buf = local_sorted.indices;
-  if (argsort && buf_size_send_idcs_total > volume) {
-    idc_send_buf = create_buffer<int64_t>(buf_size_send_idcs_total, Memory::GPU_FB_MEM);
-    size_t pos   = 0;
-    for (size_t r = 0; r < num_ranks; ++r) {
-      CHECK_CUDA(cudaMemcpyAsync(idc_send_buf.ptr(aligned_pos_idcs_send[r]),
-                                 local_sorted.indices.ptr(pos),
-                                 sizeof(int64_t) * size_send[r],
-                                 cudaMemcpyDeviceToDevice,
-                                 stream));
-      pos += size_send[r];
+    {
+      Buffer<VAL*> val_send_buffers_ptr = create_buffer<VAL*>(num_sort_ranks, Memory::Z_COPY_MEM);
+      for (size_t r = 0; r < num_sort_ranks; r++) {
+        val_send_buffers_ptr[r] = val_send_buffers[r].ptr(0);
+      }
+
+      // is either 1, 2, 4, ... , THREADS_PER_BLOCK/2
+      int segments_per_threadblock =
```

Comment:
Lines 1527-1533 seem to repeat in several places. I think you extract them out to a function.

Reply:
I extracted the launchconfig generation into a more generic utility function and added some comments.